sdk/storage_datalake/src/shared_key_authorization_policy.rs

```diff
@@ -0,0 +1,244 @@
+    )
+
+    // expected
+    // GET\n /*HTTP Verb*/
+    // \n    /*Content-Encoding*/
+    // \n    /*Content-Language*/
+    // \n    /*Content-Length (include value when zero)*/
+    // \n    /*Content-MD5*/
+    // \n    /*Content-Type*/
+    // \n    /*Date*/
+    // \n    /*If-Modified-Since */
+    // \n    /*If-Match*/
+    // \n    /*If-None-Match*/
+    // \n    /*If-Unmodified-Since*/
+    // \n    /*Range*/
+    // x-ms-date:Sun, 11 Oct 2009 21:49:13 GMT\nx-ms-version:2009-09-19\n
+    //                                  /*CanonicalizedHeaders*/
+    // /myaccount /mycontainer\ncomp:metadata\nrestype:container\ntimeout:20
+    //                                  /*CanonicalizedResource*/
+    //
+    //
+}
+
+fn add_if_exists<K: http::header::AsHeaderName>(h: &HeaderMap, key: K) -> &str {
+    match h.get(key) {
+        Some(ce) => ce.to_str().unwrap(),
+        None => "",
+    }
+}
+
+fn canonicalize_header(h: &HeaderMap) -> String {
+    let mut v_headers = h
+        .iter()
+        .filter(|(k, _v)| k.as_str().starts_with("x-ms"))
+        .map(|(k, _)| k.as_str())
+        .collect::<Vec<_>>();
+    v_headers.sort_unstable();
+
+    let mut can = String::new();
+
+    for header_name in v_headers {
+        let s = h.get(header_name).unwrap().to_str().unwrap();
+        can = can + header_name + ":" + s + "\n";
+    }
+    can
+}
+
+fn canonicalized_resource(account: &str, u: &url::Url) -> String {
+    let mut can_res: String = String::new();
+    can_res += "/";
+    can_res += account;
+
+    let paths = u.path_segments().unwrap();
+
+    for p in paths {
+        can_res.push('/');
+        can_res.push_str(&*p);
+    }
+    can_res += "\n";
+
+    // query parameters
+    let query_pairs = u.query_pairs(); //.into_owned();
+    {
+        let mut qps = Vec::new();
+        {
+            for (q, _p) in query_pairs {
+                trace!("adding to qps {:?}", q);
+
+                // add only once
+                if !(qps.iter().any(|x: &String| x == q.as_ref())) {
+                    qps.push(q.into_owned());
+                }
+            }
+        }
+
+        qps.sort();
+
+        for qparam in qps {
+            // find correct parameter
+            let ret = lexy_sort(&query_pairs, &qparam);
+
+            // debug!("adding to can_res {:?}", ret);
+
+            can_res = can_res + &qparam.to_lowercase() + ":";
+
+            for (i, item) in ret.iter().enumerate() {
+                if i > 0 {
+                    can_res += ","
+                }
+                can_res += item;
+            }
+
+            can_res += "\n";
+        }
+    };
+
+    can_res[0..can_res.len() - 1].to_owned()
+}
+
+fn lexy_sort<'a>(
+    vec: &'a url::form_urlencoded::Parse,
+    query_param: &str,
+) -> Vec<std::borrow::Cow<'a, str>> {
+    let mut v_values = Vec::new();
+
+    for item in vec.filter(|x| x.0 == *query_param) {
+        v_values.push(item.1)
+    }
```

Comment:
```suggestion
    let v_values = vec.filter(|x| x.0 == *query_param).map(|item| item.1).collect() ;
```

Reply:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.